sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs

```diff
@@ -0,0 +1,393 @@
+
+            int gap;
+            if ((gap = (int)(encryptedBlobRange.OriginalRange.Offset - encryptedBlobRange.AdjustedRange.Offset) - read) > 0)
+            {
+                plaintext.Read(new byte[gap], 0, gap);
+            }
+
+            return new LengthLimitingStream(plaintext, encryptedBlobRange.OriginalRange.Count);
+        }
+
+        private async Task<Stream> DecryptBlobAsync(Stream ciphertext, Metadata metadata,
+            EncryptedBlobRange encryptedBlobRange, bool noPadding)
+        {
+            AssertKeyAccessPresent();
+            var encryptionData = GetAndValidateEncryptionData(metadata);
+
+            byte[] IV;
+            int read = 0;
+            if (encryptedBlobRange.AdjustedRange.Offset == 0)
+            {
+                IV = encryptionData.ContentEncryptionIV;
+            }
+            else
+            {
+                IV = new byte[EncryptionConstants.ENCRYPTION_BLOCK_SIZE];
+                await ciphertext.ReadAsync(IV, 0, IV.Length).ConfigureAwait(false);
+                read = IV.Length;
+            }
+
+            var plaintext = WrapStream(ciphertext, (await GetContentEncryptionKeyAsync(encryptionData).ConfigureAwait(false)).ToArray(), encryptionData, IV, noPadding);
+
+            int gap;
+            if ((gap = (int)(encryptedBlobRange.OriginalRange.Offset - encryptedBlobRange.AdjustedRange.Offset) - read) > 0)
+            {
+                await plaintext.ReadAsync(new byte[gap], 0, gap).ConfigureAwait(false);
+            }
+
+            return new LengthLimitingStream(plaintext, encryptedBlobRange.OriginalRange.Count);
+        }
+
+        #endregion
+
+        private static Metadata ExtractMetadata(ResponseHeaders headers)
+        {
+            const string metadataPrefix = "x-ms-meta-";
+
+            Metadata metadata = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            foreach (HttpHeader header in headers)
+            {
+                if (header.Name.StartsWith(metadataPrefix, StringComparison.InvariantCulture))
+                {
+                    metadata[header.Name.Substring(metadataPrefix.Length)] = header.Value;
+                }
+            }
+
+            return metadata;
+        }
+
+        /// <summary>
+        /// Gets and validates a blob's encryption-related metadata
+        /// </summary>
+        /// <param name="metadata">The blob's metadata</param>
+        /// <returns>The relevant metadata.</returns>
+        private static EncryptionData GetAndValidateEncryptionData(Metadata metadata)
+        {
+            _ = metadata ?? throw new InvalidOperationException();
+
+            EncryptionData encryptionData;
+            if (!metadata.TryGetValue(EncryptionConstants.ENCRYPTION_DATA_KEY, out string encryptedDataString))
+            {
+                throw new InvalidOperationException("Encryption data does not exist.");
```

Comment:
Consider moving to EncryptionContants

Reply:
We are now downloading normally if data isn't encrypted. Saves customers a call if they mess up.